### ISSUE TITLE
Framework for qcert "subroutines" in Java

### ISSUE THREAD

**Issue by joshuaauerbachwatson** [querycert/qcert#28]
The original description is available as the first comment.  Here I list the tasks to be accomplished:

* [x] Create javaService component which supports an open ended set of "verbs".   Each verb names a function from a potentially long String argument to a potentially long String result.  Both forked process and HTTP server are supported.
* [x] Move the existing SQLParser service to be "verb" within javaService.
* [x] Create a parallel rule2CAMP service.   Note that for now this can only be deployed locally, so, on the public server its absence should result in a good error message and no crash.
  * [x] Get the server (also forked binary) version of the service working mostly.
  * [x] Fix issue with ARL files needing XOM information in order to translate correctly.
  * [x] Introduce appropriate call points in the qcert binary and the web form
* [ ] (Perhaps) obtain needed approvals so that rule2CAMP service can be deployed publically.
* [ ] Converge the qcert binary callout (using fork) with the web form pre-process as per the sketch in the first comment

**Comment by joshuaauerbachwatson** [comment]
The following prose used to be in the description.   The description is replaced by a task list.

Today, we have one callout from qcert to Java and it is used for SQL as a source language. We are contemplating adding at least one more for the "JRules" languages.

There is already a piece of functionality (the TableConverter) which is provided as a Java utility. This covers the functionality of encoding input and schema information into the appropriate JSON structure for ingestion either by qcert or by those of our runtimes that expect JSON input. It is not inconceivable that this could remain as a distinct utility, but this exposes the arbitrariness of our approach to schemas and data. So, doing this functionality, also, by callout is something we should consider.

In theory some of these callouts could be replaced by pure OCaml and Coq code. However, replacing any SQL support (currently dependent on Presto) would be a big task, and doing anything with JRules that way would be essentially impossible. (The data encoding, and some of the schema processing (where it doesn't require Presto) would, however, be candidates for recoding in OCaml).

So, Java callouts may be here to stay and ought to be put on a good design basis. Right now, the design is sloppy. A background to describing this sloppiness is the fact that there are two distinct ways of getting to Java code, by forking a process to run a JVM, and by making an HTTP request to a server. To support only the qcert binary, the first is more than adequate; the second exists to support running in a web browser. Standardizing on just the second way is possible but awkward since any time the system changes there is a chore of getting a new instance of the server running with the latest code. Forking has its place as a more "tightly coupled" approach where it is feasible. So, I'm assuming that the two ways should probably persist.

However,

Both the forked Java main and the server are lacking in an extensible structure because they were written just to support SQL s-expression encoding and nothing else.

Currently, the invocation site for calling out to Java is inconsistently placed. When forking a process, it is done from the qcert binary (it is a true "call out"). When going to a remote server, it is done before invoking qcert (a "pre-process" step), and causes qcert to be invoked differently because some processing has already been done. This will be a serious inhibitor to creating a design that supports extension.

The two invocation paths are used rigidly. In the binary, if there is no Java in the path or the jars have not been put in the right place, the fork fails even if there is a server running. In the browser, if the server is not running, the request fails. Furthermore, in the browser, the URL of the server is semi-hardcoded (it is either localhost or a specific remote URL and the user has to check which is to be used).

I think what we want is more like the following:

Both the server main and the co-process main should be just shells with a top level dispatch and the ability to execute more or less arbitrary code. While the shells would necessarily differ, the Java functionality in each dispatch case should be identical whether it is running in the server or in the co-process.

The "pre-process" flavor of how we do SQL encoding in the browser should be eliminated. Instead, we should have one invokeJava function in OCaml with a very loosely typed structure (basically a function name and a list of function arguments, all strings), returning a trivial variant which is either a String result or a String error message. The logic here should be set up for dual compilation, so that when compiled to binary, this code will first try to fork a process, then, if that fails, will follow some configurable path of possible URLs until it finds one that responds. When compiled to Javascript, hand-written Javascript would entirely replace the function so that no fork is attempted and the same path of URLs is followed, this time using the browser DOM appropriately to avoid sandbox violations. Inside qcert, it would have to be known what the legal repertoire of possible functions and arguments are (perhaps calls to the loosely typed function could be funneled through a more strongly typed one).

Everything else would be written once. The invocation side of each callout would be written once, going through the invokeJava choke point. The implementation side would be written once as a Java class that can be invoked identically from both the process shell and the server shell.

**Comment by joshuaauerbachwatson** [comment]
(The following is not yet pushed, pending resolution of the problems).

I actually created two new verbs in the JavaService, one (techRule2CAMP) is for ARL source, and the other (serialRule2CAMP) is for serialized SemRule trees.   Testing of the first of these fails because the parsing of ARL rules expects the types employed to be in the semantic object model.  In our usual approach to testing ARL rules, we rely on the fact that the needed types are already in the XOM (that is, as Java classes in the classpath) and these get incorporated into the semantic object model on demand.   

There is an alternative strategy whereby the types that are employed can be listed and placed in the semantic object model explicitly.   I have to see whether that works, and then decide how to shoehorn the required list of types through the limited JavaService text interface.   Ultimately, this will depend on how the separation of schema, input, and output in the I/O file shakes out (as per what @jeromesimeon is doing).    What is needed here is the source plus the schema (no input or output).   I am inclined to try something expedient and equivalent for now until we understand how the separation works.

Testing of the second verb is not yet underway, also due to the same limited text interface.  In this case, we can't put arbitrary binary (such as is found in a .sem file) through that interface so the plan is to use base64 basic encoding.  As soon as that's plugged in (pretty simple) I will have more info on where we stand.

**Comment by joshuaauerbachwatson** [comment]
Update on where we stand:

- For .sem files, the server "just works" but of course it is only doing translation.  The problem with these files is that we are not really sure what semantics to assign to different rule dialects or how to do the execution phase

- For .arl files, the idea of using a list of types does _not_ work and I will back that out before pushing.  As of the expected next push, .arl rules will not build in the integrated code.   In the local case, we (or even an end user) can pre-deploy a jar containing the XOM classes.   This obviously would never work in a remote server, so we would have to send the jar over the network as part of the http post; while that sounds crazy I think it would work.

**Comment by joshuaauerbachwatson** [comment]
While a simple list of types doesn't work for ARL, a complete schema (including attributes of the classes) works, if properly encoded.  I solved the problem by consuming a schema (basically the "schema" section of the new I/O file format), and sending it along with the source.  It gets turned into the appropriate SemObjectModel form at the server.   ARL files now work in the "integrated" code

To be clear, the integration isn't complete (see task list).  But I am able to test the javaService end by emulating the insertion points in the qcert binary and the web form using a tester.  And that is all working.

**Comment by joshuaauerbachwatson** [comment]
Of the incomplete tasks in this issue at this point:

"introduce appropriate call points in the qcert binary and the web form" (subtask specifically related to JRules): this is done for the web form.  For the qcert binary, between @jeromesimeon or myself we should just do this and close this issue.
"(Perhaps) obtain needed approvals so that rule2CAMP service can be deployed publically":  the "perhaps" says it all.  If this happens great, but I won't hold up closing the issue for it.
"Converge the qcert binary callout (using fork) with the web form pre-process as per the sketch in the first comment":  I don't think we need to do this now.  At some point downstream, if we feel we need added elegance to counteract creeping chaos, we can do it then.

**Comment by joshuaauerbachwatson** [comment]
I have checked off the task for the jrules2CAMP service because I've achieved a working integration into the qcert binary.  Of course, I left some hacks in CompDriver.v.  I will open a narrowly focused issue to clean that up and assign it to @jeromesimeon (sorry Jerome).

As per the previous comment, the remaining two tasks are being left undone.  

If we think the rule2CAMP service is worth deploying publically, we need Martin to pursue that (probably).  I'm not opening an issue for it, both because I'm not sure how critical it is, and because Martin is not listed in querycert people, so I couldn't assign the issue to him anyway.   Note that even without public deployment, open source users have enough to run their own private local service (or use the qcert binary) as long as they have a legal copy of an ODM Designer product.

The final tighter integration task (coalescing the qcert.html way of doing callouts and the qcert binary ocaml way) is just being dropped since I think what we have is good enough for now.
